File: src/TacheA.c


#include "TacheA.h"

int csub(char x_i, char y_i)
{
  printf("\n[csub %c %c] ...", x_i, y_i);
  if (x_i == y_i)
  {
    DEBUG();
    printf("\n[csub %c %c] 0 ...ok", x_i, y_i);

    return 0;
  }
  if ((x_i == 'A' && y_i == 'T') || (x_i == 'G' && y_i == 'C') || (x_i == 'C' && y_i == 'G') || (x_i == 'T' && y_i == 'A'))
  {
    DEBUG();
    printf("\n[csub %c %c] 3 ...ok", x_i, y_i);

    return 3;
  }
  printf("\n[csub %c %c] 4 ...ok", x_i, y_i);
  return 4;
}




int DIST_NAIF_REC (char** x, char** y, int i, int j, int c, int dist)
{
  // Enumerer tous les alignements possibles de deux mots x et y
  printf("\n[DIST_NAIF_REC %s %s %d %d %d %d] ...", *x, *y, i, j, c, dist);
  // sleep(1);
  // printf("\n*x[0] = %c", x[0][0]);
  // printf("\n*x[1] = %c", x[0][1]);
  // printf("\n*x[2] = %c", x[0][2]);
  // printf("\n*x[3] = %c", *(x[strlen(*x - 1)]));
  printf("\nstrlen(*x) = %ld", strlen(*x));

  DEBUG();
  if (i>=strlen(x[0]) && j>=strlen(y[0]) )
  {
    printf("\n/!/ i = %d, j = %d", i, j);
    DEBUG();
    if (dist == INFINI || c < dist)
    {
      DEBUG();
      dist = c;
    }
  }
  else
  {
    DEBUG();
    if (i < strlen(*x) && j < strlen(*y))
    {
      printf("\nCas 1");
      printf("\ni = %d, j = %d", i, j);
      DEBUG();
      dist = DIST_NAIF_REC(x, y, i+1, j+1, c + csub(x[0][i], y[0][j]), dist );
    }
    if (i < strlen(*x))
    {
      printf("\nCas 2 : i=%d < strlen(*x)=%ld", i, strlen(*x));

      DEBUG();
      dist = DIST_NAIF_REC(x, y, i+1, j, c + C_DEL, dist);
    }
    if (j < strlen(*y))
    {
      printf("\nCas 3");

      DEBUG();
      dist = DIST_NAIF_REC(x, y, i, j+1, c + C_INS, dist);
    }

  }
  printf("\n[DIST_NAIF_REC %s %s %d %d %d %d] ...ok", *x, *y, i, j, c, dist);

  return dist;
}

int DIST_NAIF(char** x, char** y)
{

  return DIST_NAIF_REC(x, y, 0, 0, 0, INFINI);
}


int main(int argc, char* argv[])
{
  if (argc == 3)
  {
    // printf("\n%d\n", 'A');
    char* arg1 = argv[1];
    char* arg2 = argv[2];
    printf("\n\nDIST_NAIF(%s, %s) = %d\n\n", arg1, arg2, DIST_NAIF(&arg1, &arg2) );
    return 0;
  }
  if (argc == 2)
  {
    int n;
    int m;
    char* x;
    char* y;
    if (read_adnFILE(argv[1], &n, &m, &x, &y) != 0)
    {
      printf("\nErreur lors de la lecture d'un adnFile");
      return 0;
    }
    printf("\nn = %d, m = %d, x = %s, y = %s\n", n, m, x, y);

    return 0;
  }
  printf("\nErreur : nombre d'arguments invalide\n");
  return 0;
}

File: src/TacheA.py
#!/usr/bin/python3.6.7
import glob
import time
#from memory_profiler import memory_usage

cdel = 2
cins = 2
INFINI = -1


def csub(x_i, y_i) :
    """
    char * char -> int
    Retourne le cout de la substitution (x_i, y_i).
    """
    if (x_i == y_i) :
        return 0
    if ((x_i == 'A' and y_i == 'T') or (x_i == 'G' and y_i == 'C') or (x_i == 'C' and y_i == 'G')or (x_i == 'T' and y_i == 'A')) :
        return 3
    return 4


def DIST_NAIF(x, y) :
    """
    str*str->int
    """
    return DIST_NAIF_REC(x, y, 0, 0, 0, INFINI)

def DIST_NAIF_REC (x, y, i, j, c, dist) :
    """
    str*str*int*int*int*int -> int
    Enumerer tous les alignements possibles de deux mots x et y
    """
    #print("DIST_NAIF_REC(" + str(x) + ", "  + str(y) + ", "+ str(i) + ", "  + str(j) + ", "  + str(c) + ", " + str(dist)+")")

    if (i==len(x) and j==len(y) ) :
        if (dist == INFINI or c < dist) :
            dist = c
    else :
        if (i < len(x) and j < len(y)) :
            dist = DIST_NAIF_REC(x, y, i+1, j+1, c + csub(x[i], y[j]), dist )
        if (i < len(x)) :
            dist = DIST_NAIF_REC(x, y, i+1, j, c + cdel, dist)
        if (j < len(y)) :
            dist = DIST_NAIF_REC(x, y, i, j+1, c + cins, dist)
    #print(dist)
    return dist




#file_object = open(<str : filename>, <str : mode : r / w / a / r+>)

def dist_on_file(filename) :
    """
    str -> int
    Lit deux mots x et y a partir d'un adn file, et retourne la distance entre les deux.
    """
    f = open(filename, 'r')
    x_size = int(f.readline())
    y_size = int(f.readline())
    x = (f.readline()).replace(' ', '')
    y = (f.readline()).replace(' ', '')
    print (x + "" + y, end='')

    f.close()

    return DIST_NAIF_REC(x, y, 0, 0, 0, INFINI)

def test() :

    assert DIST_NAIF_REC('ATTGTA', 'ATCTTA', 0, 0, 0, INFINI) == 4


    assert dist_on_file("../Instances/Inst_0000010_44.adn") == 10
    assert dist_on_file("../Instances/Inst_0000010_7.adn") == 8

    assert dist_on_file("../Instances/Inst_0000010_8.adn") == 2

# 0.3% de memoire utilisee, 100% de la RAM
def testminute() :

    dir_list = glob.glob("../Instances/*.adn")
    dir_list.sort()
    #print (dir_list)

    for f in dir_list :
        print ("_" * len(f) + "\n" + f)
        start_time = time.time()
        dist_on_file(f)
        elapsed_time = time.time() - start_time
        estimate = round(elapsed_time, 2)
        if elapsed_time > 60 :
            min = str(estimate//60)
            sec = str(estimate%60)
            print ("Temps d'execution : " + min + "m " + sec + "s\n")
            return


        print ("Temps d'execution : " + str(estimate) + "s\n")
